Remotion/Web/Development.WebTesting/Utilities/ProcessUtils.cs

```diff
@@ -134,16 +136,103 @@ public static void GracefulProcessShutdown ([NotNull] Process process, TimeSpan
       }
-      
+
       // Try to gracefully close the process
-      if (process.CloseMainWindow())
+      if (CloseMainWindow (process))
       {
         // No exception is thrown when calling .WaitForExit(..) on an already exited process.
         if (process.WaitForExit (timeoutInMilliseconds))
           return;
       }
 
       // Force closing the process
+      KillProcess (process);
+
+      // No exception is thrown when calling .WaitForExit(..) on an already exited process.
+      if (process.WaitForExit (timeoutInMilliseconds))
+        return;
+
+      throw CreateProcessDidNotExitInTimeException (process);
+    }
+
+    /// <summary>
+    /// Tries to shutdown a list of processes gracefully and if they are not closed after <paramref name="timeout"/> it ensures that they are closed.
+    /// </summary>
+    /// <param name="processes">The processes that should be shut down.</param>
+    /// <param name="timeout">The wait timeout after a shutdown has been requested.</param>
+    /// <remarks>
+    /// <p>
+    /// This method will block until the processes have exited. Shutdown procedure is as follows:
+    /// </p>
+    /// <list type="number">
+    /// <item>
+    /// <description>Check if the processes have already exited.</description>
+    /// </item>
+    /// <item>
+    /// <description>Close the main windows of the processes and wait for the processes to exit or the <paramref name="timeout"/> to run out.</description>
+    /// </item>
+    /// <item>
+    /// <description>If the processes are still not closed, kill them and wait for the processes to exit or the <paramref name="timeout"/> to run out.</description>
+    /// </item>
+    /// </list>
+    /// </remarks>
+    public static void GracefulProcessShutdown (IReadOnlyList<Process> processes, TimeSpan timeout)
+    {
+      ArgumentUtility.CheckNotNullOrItemsNull (nameof (processes), processes);
+
+      var timeoutInMilliseconds = (int) timeout.TotalMilliseconds;
+      if (timeoutInMilliseconds < 0)
+        throw new ArgumentOutOfRangeException ("timeout", "Timeout can not be smaller that zero.");
+
+      IReadOnlyList<Process> remainingProcesses = processes.ToList();
+
+      // Clear out any processes that already exited
+      remainingProcesses = WaitForProcessesExits (remainingProcesses, TimeSpan.Zero);
+
+      // Try to gracefully close the process
+      var anyMainWindowClosed = remainingProcesses.Any (CloseMainWindow);
+      if (anyMainWindowClosed)
+        remainingProcesses = WaitForProcessesExits (remainingProcesses, timeout);
+
+      // Force closing the process and wait for process exits
+      foreach (var remainingProcess in remainingProcesses)
+        KillProcess (remainingProcess);
+
+      remainingProcesses = WaitForProcessesExits (remainingProcesses, timeout);
+
+      if (remainingProcesses.Count == 0)
```

Comment:
Would this be cleaner if we inverted to the if and just had the exception in the conditonal and otherwise just a regular method exit? We could then also say remainingProcesses.Any() in the IF.